File: Icon.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2009, 2010, 2011, 2012 Sidoine De Wispelaere.
    Copyright (C) 2012, 2013, 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

local OVALE, Ovale = ...

--<class name="OvaleIcon" inherits="ActionButtonTemplate" />

--<private-static-properties>
local L = Ovale.L
local OvaleSpellBook = Ovale.OvaleSpellBook
local OvaleState = Ovale.OvaleState

local format = string.format
local next = next
local pairs = pairs
local strfind = string.find
local strsub = string.sub
local tostring = tostring
local API_GetTime = GetTime
local API_PlaySoundFile = PlaySoundFile
local INFINITY = math.huge
-- GLOBALS: _G
-- GLOBALS: GameTooltip

-- Threshold for resetting the cooldown animation.
-- This should be more than OvaleFuture's SIMULATOR_LAG.
local COOLDOWN_THRESHOLD = 0.1
--</private-static-properties>

local function HasScriptControls()
	return (next(Ovale.checkBoxWidget) ~= nil or next(Ovale.listWidget) ~= nil)
end

--<public-methods>
local function SetValue(self, value, actionTexture)
	self.icone:Show()
	self.icone:SetTexture(actionTexture)
	self.icone:SetAlpha(Ovale.db.profile.apparence.alpha)
	self.cd:Hide()
	self.focusText:Hide()
	self.rangeIndicator:Hide()
	self.shortcut:Hide()
	if value then
		self.actionType = "value"
		self.actionHelp = nil
		self.value = value
		if value < 10 then
			self.remains:SetFormattedText("%.1f", value)
		elseif value == INFINITY then
			self.remains:SetFormattedText("inf")
		else
			self.remains:SetFormattedText("%d", value)
		end
		self.remains:Show()
	else
		self.remains:Hide()
	end
	self:Show()
end

local function Update(self, element, startTime, actionTexture, actionInRange, actionCooldownStart, actionCooldownDuration,
				actionUsable, actionShortcut, actionIsCurrent, actionEnable, actionType, actionId, actionTarget, actionResourceExtend)
	self.actionType = actionType
	self.actionId = actionId
	self.value = nil

	local now = API_GetTime()
	local state = OvaleState.state
	local profile = Ovale.db.profile

	if startTime and actionTexture then
		-- Cooldown.
		local cd = self.cd
		local resetCooldown = false
		if startTime > now then
			--[[
				Check for the cooldown duration mysteriously being reset to zero while an active
				cooldown animation is in effect, and force the cooldown information to be reset
				to the existing information.
			--]]
			local duration = cd:GetCooldownDuration()
			if duration == 0 and self.texture == actionTexture and self.cooldownStart and self.cooldownEnd then
				resetCooldown = true
			end
			if self.texture ~= actionTexture or not self.cooldownStart or not self.cooldownEnd then
				-- Update the cooldown if this is a new action
				self.cooldownStart = now
				self.cooldownEnd = startTime
				resetCooldown = true
			elseif startTime < self.cooldownEnd - COOLDOWN_THRESHOLD
					or startTime > self.cooldownEnd + COOLDOWN_THRESHOLD then
				-- Update the cooldown if the new end time would exceed the tolerance threshold.
				if startTime - self.cooldownEnd > 0.25 or startTime - self.cooldownEnd < -0.25 then 
					self.cooldownStart = now
				else
					--[[ 
						Modify the start and end time of the cooldown swipe so that it maintains the same 
						percent progress and only accelerates/decelerates the swipe.
					]]--
					local oldCooldownProgressPercent = (now - self.cooldownStart) / (self.cooldownEnd - self.cooldownStart)
					self.cooldownStart = (now - oldCooldownProgressPercent * startTime) / (1 - oldCooldownProgressPercent)
				end
				self.cooldownEnd = startTime
				resetCooldown = true
			end
			self.texture = actionTexture
		else
			-- The cooldown is finished.
			self.cooldownStart = nil
			self.cooldownEnd = nil
		end
		if self.cdShown and profile.apparence.flashIcon and self.cooldownStart and self.cooldownEnd then
			local start, ending = self.cooldownStart, self.cooldownEnd
			local duration = ending - start
			if resetCooldown and duration > COOLDOWN_THRESHOLD then
				cd:SetDrawEdge(false)
				cd:SetSwipeColor(0, 0, 0, 0.8)
				cd:SetCooldown(start, duration)
				cd:Show()
			end
		else
			self.cd:Hide()
		end

		-- Icon texture and cooldown
		self.icone:Show()
		self.icone:SetTexture(actionTexture)

		local alpha = profile.apparence.alpha
		if actionUsable then
			self.icone:SetAlpha(alpha)
		else
			-- Unusable actions are displayed at half the normal alpha.
			alpha = alpha / 2
			self.icone:SetAlpha(alpha)
		end

		-- Icon color overlay (red or not red).
		if element.namedParams.nored ~= 1 and actionResourceExtend and actionResourceExtend > 0 then
			self.icone:SetVertexColor(0.75, 0.2, 0.2)
		else
			self.icone:SetVertexColor(1, 1, 1)
		end

		-- Action help text.
		self.actionHelp = element.namedParams.help

		-- Sound file.
		if not (self.cooldownStart and self.cooldownEnd) then
			self.lastSound = nil
		end
		if element.namedParams.sound and not self.lastSound then
			local delay = element.namedParams.soundtime or 0.5
			if now >= startTime - delay then
				self.lastSound = element.namedParams.sound
			--	print("Play" .. self.lastSound)
				API_PlaySoundFile(self.lastSound)
			end
		end

		if not red and startTime > now and profile.apparence.highlightIcon then
			local lag = 0.6
			local newShouldClick = (startTime < now + lag)
			if self.shouldClick ~= newShouldClick then
				if newShouldClick then
					self:SetChecked(true)
				else
					self:SetChecked(false)
				end
				self.shouldClick = newShouldClick
			end
		elseif self.shouldClick then
			self.shouldClick = false
			self:SetChecked(false)
		end

		-- Remaining time.
		if (profile.apparence.numeric or self.namedParams.text == "always") and startTime > now then
			self.remains:SetFormattedText("%.1f", startTime - now)
			self.remains:Show()
		else
			self.remains:Hide()
		end

		-- Keyboard shortcut.
		if profile.apparence.raccourcis then
			self.shortcut:Show()
			self.shortcut:SetText(actionShortcut)
		else
			self.shortcut:Hide()
		end

		-- Range indicator.
		if actionInRange == 1 then
			self.rangeIndicator:SetVertexColor(0.6,0.6,0.6)
			self.rangeIndicator:Show()
		elseif actionInRange == 0 then
			self.rangeIndicator:SetVertexColor(1.0,0.1,0.1)
			self.rangeIndicator:Show()
		else
			self.rangeIndicator:Hide()
		end

		-- Focus text.
		if element.namedParams.text then
			self.focusText:SetText(tostring(element.namedParams.text))
			self.focusText:Show()
		elseif actionTarget and actionTarget ~= "target" then
			self.focusText:SetText(actionTarget)
			self.focusText:Show()
		else
			self.focusText:Hide()
		end

		self:Show()
	else
		self.icone:Hide()
		self.rangeIndicator:Hide()
		self.shortcut:Hide()
		self.remains:Hide()
		self.focusText:Hide()
		if profile.apparence.hideEmpty then
			self:Hide()
		else
			self:Show()
		end
		if self.shouldClick then
			self:SetChecked(false)
			self.shouldClick = false
		end
	end

	return startTime, element
end

local function SetHelp(self, help)
	self.help = help
end

local function SetParams(self, positionalParams, namedParams, secure)
	self.positionalParams = positionalParams
	self.namedParams = namedParams

	self.actionButton = false
	if secure then
		for k, v in pairs(namedParams) do
			local index = strfind(k, "spell")
			if index then
				local prefix = strsub(k, 1, index - 1)
				local suffix = strsub(k, index + 5)
				self:SetAttribute(prefix .. "type" .. suffix, "spell")
				self:SetAttribute("unit", self.namedParams.target or "target")
				self:SetAttribute(k, OvaleSpellBook:GetSpellName(v))
				self.actionButton = true
			end
		end
	end
end

local function SetRemainsFont(self, color)
	self.remains:SetTextColor(color.r, color.g, color.b, 1.0)
	self.remains:SetJustifyH("left")
	self.remains:SetPoint("BOTTOMLEFT", 2, 2)	
end

local function SetFontScale(self, scale)
	self.fontScale = scale
	self.remains:SetFont(self.fontName, self.fontHeight * self.fontScale, self.fontFlags)
	self.shortcut:SetFont(self.fontName, self.fontHeight * self.fontScale, self.fontFlags)
	self.rangeIndicator:SetFont(self.fontName, self.fontHeight * self.fontScale, self.fontFlags)
	self.focusText:SetFont(self.fontName, self.fontHeight * self.fontScale, self.fontFlags)
end

local function SetRangeIndicator(self, text)
	self.rangeIndicator:SetText(text)
end
--</public-methods>

local function OvaleIcon_OnMouseUp(self)
	if not self.actionButton then
		Ovale:ToggleOptions()
	end
	self:SetChecked(true)
end

-- GLOBALS: OvaleIcon_OnEnter
function OvaleIcon_OnEnter(self)
	if self.help or self.actionType or HasScriptControls() then
		GameTooltip:SetOwner(self, "ANCHOR_BOTTOMLEFT")
		if self.help then
			GameTooltip:SetText(L[self.help])
		end
		if self.actionType then
			local actionHelp = self.actionHelp
			if not actionHelp then
				if self.actionType == "spell" then
					actionHelp = OvaleSpellBook:GetSpellName(self.actionId) -- .. " (" .. self.actionId .. ")"
				elseif self.actionType == "value" then
					actionHelp = (self.value < INFINITY) and tostring(self.value) or "infinity"
				else
					actionHelp = format("%s %s", self.actionType, tostring(self.actionId))
				end
			end
			GameTooltip:AddLine(actionHelp, 0.5, 1, 0.75)
		end
		if HasScriptControls() then
			GameTooltip:AddLine(L["Cliquer pour afficher/cacher les options"], 1, 1, 1)
		end
		GameTooltip:Show()
	end
end

-- GLOBALS: OvaleIcon_OnLeave
function OvaleIcon_OnLeave(self)
	if self.help or HasScriptControls() then
		GameTooltip:Hide()
	end
end

-- GLOBALS: OvaleIcon_OnLoad
function OvaleIcon_OnLoad(self)
	local name = self:GetName()
	local profile = Ovale.db.profile

--<public-properties>
	self.icone = _G[name.."Icon"]
	self.shortcut = _G[name.."HotKey"]
	self.remains = _G[name.."Name"]
	self.rangeIndicator = _G[name.."Count"]
	self.rangeIndicator:SetText(profile.apparence.targetText)
	self.cd = _G[name.."Cooldown"]
	self.normalTexture = _G[name.."NormalTexture"]
	local fontName, fontHeight, fontFlags = self.shortcut:GetFont()
	self.fontName = fontName
	self.fontHeight = fontHeight
	self.fontFlags = fontFlags
	self.focusText = self:CreateFontString(nil, "OVERLAY");
	self.cdShown = true
	self.shouldClick = false
	self.help = nil
	self.value = nil
	self.fontScale = nil
	self.lastSound = nil
	self.cooldownEnd = nil
	self.cooldownStart = nil
	self.texture = nil
	self.positionalParams = nil
	self.namedParams = nil
	self.actionButton = false
	self.actionType = nil
	self.actionId = nil
	self.actionHelp = nil
--</public-properties>

	self:SetScript("OnMouseUp", OvaleIcon_OnMouseUp)

	self.focusText:SetFontObject("GameFontNormalSmall");
	self.focusText:SetAllPoints(self);
	self.focusText:SetTextColor(1,1,1);
	self.focusText:SetText(L["Focus"])

	--self:RegisterForClicks("LeftButtonUp")
	self:RegisterForClicks("AnyUp")
	self.Update = Update
	self.SetHelp = SetHelp
	self.SetParams = SetParams
	self.SetRemainsFont = SetRemainsFont
	self.SetFontScale = SetFontScale
	self.SetRangeIndicator = SetRangeIndicator
	self.SetValue = SetValue
	if profile.clickThru then
		self:EnableMouse(false)
	end
end

File: src/Icon.ts
import { L } from "./Localization";
import { format, find, sub } from "@wowts/string";
import { next, tostring, _G, kpairs } from "@wowts/lua";
import { GetTime, PlaySoundFile, UIFrame, UIFontString, UITexture, UICooldown, UICheckButton, CreateFrame, GameTooltip, UIPosition } from "@wowts/wow-mock";
import { huge } from "@wowts/math";
import { NamedParameters, PositionalParameters, AstNode } from "./AST";
import { OvaleOptionsClass } from "./Options";
import { OvaleSpellBookClass } from "./SpellBook";
let INFINITY = huge;
let COOLDOWN_THRESHOLD = 0.1;

interface IconParent {
    checkBoxWidget: any;
    listWidget: any;
    frame: UIFrame;
    ToggleOptions(): void; 
}

export class OvaleIcon {
    actionHelp: string | undefined;
    actionId: any;
    actionType: any;
    actionButton: boolean = false;
    namedParams: any;
    positionalParams: any;
    texture: string | undefined;
    cooldownStart: any;
    cooldownEnd: any;
    lastSound: any;
    fontScale: any;
    value: any;
    help: string | undefined;
    shouldClick: boolean = false;
    cdShown: boolean = false;
    focusText: UIFontString;
    fontFlags: any;
    fontHeight: any;
    fontName: any;
    normalTexture: UITexture;
    cd: UICooldown;
    rangeIndicator: UIFontString;
    remains: UIFontString;
    shortcut: UIFontString;
    icone: UITexture;
    frame: UICheckButton;

    HasScriptControls() {
        return (next(this.parent.checkBoxWidget) != undefined || next(this.parent.listWidget) != undefined);
    }

    constructor(name: string, private parent: IconParent, secure: boolean, private ovaleOptions: OvaleOptionsClass, private ovaleSpellBook: OvaleSpellBookClass) {
        if (!secure) {
            this.frame = CreateFrame("CheckButton", name, parent.frame, "ActionButtonTemplate");
        }        
        else{
            this.frame = CreateFrame("CheckButton", name, parent.frame, "SecureActionButtonTemplate, ActionButtonTemplate");
        }
        const profile = this.ovaleOptions.db.profile;
        this.icone = _G[`${name}Icon`];
        this.shortcut = _G[`${name}HotKey`];
        this.remains = _G[`${name}Name`];
        this.rangeIndicator = _G[`${name}Count`];
        this.rangeIndicator.SetText(profile.apparence.targetText);
        this.cd = _G[`${name}Cooldown`];
        this.normalTexture = _G[`${name}NormalTexture`];
        let [fontName, fontHeight, fontFlags] = this.shortcut.GetFont();
        this.fontName = fontName;
        this.fontHeight = fontHeight;
        this.fontFlags = fontFlags;
        this.focusText = this.frame.CreateFontString(undefined, "OVERLAY");
        this.cdShown = true;
        this.shouldClick = false;
        this.help = undefined;
        this.value = undefined;
        this.fontScale = undefined;
        this.lastSound = undefined;
        this.cooldownEnd = undefined;
        this.cooldownStart = undefined;
        this.texture = undefined;
        this.positionalParams = undefined;
        this.namedParams = undefined;
        this.actionButton = false;
        this.actionType = undefined;
        this.actionId = undefined;
        this.actionHelp = undefined;
        this.frame.SetScript("OnMouseUp", () => this.OvaleIcon_OnMouseUp());
        this.frame.SetScript("OnEnter", () => this.OvaleIcon_OnEnter());
        this.frame.SetScript("OnLeave", () => this.OvaleIcon_OnLeave());
        this.focusText.SetFontObject("GameFontNormalSmall");
        this.focusText.SetAllPoints(this.frame);
        this.focusText.SetTextColor(1, 1, 1);
        this.focusText.SetText(L["Focus"]);
        this.frame.RegisterForClicks("AnyUp");
        if (profile.apparence.clickThru) {
            this.frame.EnableMouse(false);
        }
    }

    SetValue(value: number | undefined, actionTexture: string | undefined) {
        this.icone.Show();
        this.icone.SetTexture(actionTexture);
        this.icone.SetAlpha(this.ovaleOptions.db.profile.apparence.alpha);
        this.cd.Hide();
        this.focusText.Hide();
        this.rangeIndicator.Hide();
        this.shortcut.Hide();
        if (value) {
            this.actionType = "value";
            this.actionHelp = undefined;
            this.value = value;
            if (value < 10) {
                this.remains.SetFormattedText("%.1f", value);
            } else if (value == INFINITY) {
                this.remains.SetFormattedText("inf");
            } else {
                this.remains.SetFormattedText("%d", value);
            }
            this.remains.Show();
        } else {
            this.remains.Hide();
        }
        this.frame.Show();
    }
    Update(element?: AstNode, startTime?: number, actionTexture?: string, actionInRange?: boolean, actionCooldownStart?: number, actionCooldownDuration?: number, actionUsable?: boolean, actionShortcut?: string, actionIsCurrent?: boolean, actionEnable?: boolean, actionType?: string, actionId?: string | number, actionTarget?: string, actionResourceExtend?: number) {
        this.actionType = actionType;
        this.actionId = actionId;
        this.value = undefined;
        let now = GetTime();
        const profile = this.ovaleOptions.db.profile;
        if (startTime && actionTexture) {
            let cd = this.cd;
            let resetCooldown = false;
            if (startTime > now) {
                let duration = cd.GetCooldownDuration();
                if (duration == 0 && this.texture == actionTexture && this.cooldownStart && this.cooldownEnd) {
                    resetCooldown = true;
                }
                if (this.texture != actionTexture || !this.cooldownStart || !this.cooldownEnd) {
                    this.cooldownStart = now;
                    this.cooldownEnd = startTime;
                    resetCooldown = true;
                } else if (startTime < this.cooldownEnd - COOLDOWN_THRESHOLD || startTime > this.cooldownEnd + COOLDOWN_THRESHOLD) {
                    if (startTime - this.cooldownEnd > 0.25 || startTime - this.cooldownEnd < -0.25) {
                        this.cooldownStart = now;
                    } else {
                        let oldCooldownProgressPercent = (now - this.cooldownStart) / (this.cooldownEnd - this.cooldownStart);
                        this.cooldownStart = (now - oldCooldownProgressPercent * startTime) / (1 - oldCooldownProgressPercent);
                    }
                    this.cooldownEnd = startTime;
                    resetCooldown = true;
                }
                this.texture = actionTexture;
            } else {
                this.cooldownStart = undefined;
                this.cooldownEnd = undefined;
            }
            if (this.cdShown && profile.apparence.flashIcon && this.cooldownStart && this.cooldownEnd) {
                let [start, ending] = [this.cooldownStart, this.cooldownEnd];
                let duration = ending - start;
                if (resetCooldown && duration > COOLDOWN_THRESHOLD) {
                    cd.SetDrawEdge(false);
                    cd.SetSwipeColor(0, 0, 0, 0.8);
                    cd.SetCooldown(start, duration);
                    cd.Show();
                }
            } else {
                this.cd.Hide();
            }
            this.icone.Show();
            this.icone.SetTexture(actionTexture);
            if (actionUsable) {
                this.icone.SetAlpha(1);
            } else {
                this.icone.SetAlpha(0.5);
            }

            if (element) {
                if (element.namedParams.nored != 1 && actionResourceExtend && actionResourceExtend > 0) {
                    this.icone.SetVertexColor(0.75, 0.2, 0.2);
                } else {
                    this.icone.SetVertexColor(1, 1, 1);
                }
                this.actionHelp = element.namedParams.help;
                if (!(this.cooldownStart && this.cooldownEnd)) {
                    this.lastSound = undefined;
                }
                if (element.namedParams.sound && !this.lastSound) {
                    let delay = element.namedParams.soundtime || 0.5;
                    if (now >= startTime - delay) {
                        this.lastSound = element.namedParams.sound;
                        PlaySoundFile(this.lastSound);
                    }
                }
            }

            let red = false; // TODO This value is not set anymore, find why
            if (!red && startTime > now && profile.apparence.highlightIcon) {
                let lag = 0.6;
                let newShouldClick = (startTime < now + lag);
                if (this.shouldClick != newShouldClick) {
                    if (newShouldClick) {
                        this.frame.SetChecked(true);
                    } else {
                        this.frame.SetChecked(false);
                    }
                    this.shouldClick = newShouldClick;
                }
            } else if (this.shouldClick) {
                this.shouldClick = false;
                this.frame.SetChecked(false);
            }
            if ((profile.apparence.numeric || this.namedParams.text == "always") && startTime > now) {
                this.remains.SetFormattedText("%.1f", startTime - now);
                this.remains.Show();
            } else {
                this.remains.Hide();
            }
            if (profile.apparence.raccourcis) {
                this.shortcut.Show();
                this.shortcut.SetText(actionShortcut);
            } else {
                this.shortcut.Hide();
            }
            if (actionInRange === undefined) {
                this.rangeIndicator.Hide();   
            } else if (actionInRange) {
                this.rangeIndicator.SetVertexColor(0.6, 0.6, 0.6);
                this.rangeIndicator.Show();
            } else {
                this.rangeIndicator.SetVertexColor(1.0, 0.1, 0.1);
                this.rangeIndicator.Show();
            } 
            if (element && element.namedParams.text) {
                this.focusText.SetText(tostring(element.namedParams.text));
                this.focusText.Show();
            } else if (actionTarget && actionTarget != "target") {
                this.focusText.SetText(actionTarget);
                this.focusText.Show();
            } else {
                this.focusText.Hide();
            }
            this.frame.Show();
        } else {
            this.icone.Hide();
            this.rangeIndicator.Hide();
            this.shortcut.Hide();
            this.remains.Hide();
            this.focusText.Hide();
            if (profile.apparence.hideEmpty) {
                this.frame.Hide();
            } else {
                this.frame.Show();
            }
            if (this.shouldClick) {
                this.frame.SetChecked(false);
                this.shouldClick = false;
            }
        }
        return [startTime, element];
    }
    SetHelp(help: string | undefined) {
        this.help = help;
    }
    SetParams(positionalParams: PositionalParameters, namedParams: NamedParameters, secure?: boolean) {
        this.positionalParams = positionalParams;
        this.namedParams = namedParams;
        this.actionButton = false;
        if (secure) {
            for (const [k, v] of kpairs(namedParams)) {
                let [index] = find(k, "spell");
                if (index) {
                    let prefix = sub(k, 1, index - 1);
                    let suffix = sub(k, index + 5);
                    this.frame.SetAttribute(`${prefix}type${suffix}`, "spell");
                    this.frame.SetAttribute("unit", this.namedParams.target || "target");
                    this.frame.SetAttribute(k, this.ovaleSpellBook.GetSpellName(<number>v) || "Unknown spell");
                    this.actionButton = true;
                }
            }
        }
    }
    SetRemainsFont(color: {r: number, g: number; b: number }) {
        this.remains.SetTextColor(color.r, color.g, color.b, 1.0);
        this.remains.SetJustifyH("left");
        this.remains.SetPoint("BOTTOMLEFT", 2, 2);
    }
    SetFontScale(scale: number) {
        this.fontScale = scale;
        this.remains.SetFont(this.fontName, this.fontHeight * this.fontScale, this.fontFlags);
        this.shortcut.SetFont(this.fontName, this.fontHeight * this.fontScale, this.fontFlags);
        this.rangeIndicator.SetFont(this.fontName, this.fontHeight * this.fontScale, this.fontFlags);
        this.focusText.SetFont(this.fontName, this.fontHeight * this.fontScale, this.fontFlags);
    }
    SetRangeIndicator(text: string) {
        this.rangeIndicator.SetText(text);
    }
    OvaleIcon_OnMouseUp() {
        if (!this.actionButton) {
            this.parent.ToggleOptions();
        }
        this.frame.SetChecked(true);
    }
    OvaleIcon_OnEnter() {
        if (this.help || this.actionType || this.HasScriptControls()) {
            GameTooltip.SetOwner(this.frame, "ANCHOR_BOTTOMLEFT");
            if (this.help) {
                GameTooltip.SetText(L[this.help]);
            }
            if (this.actionType) {
                let actionHelp: string;
                if (this.actionHelp) {
                    actionHelp = this.actionHelp
                }
                else {
                    if (this.actionType == "spell") {
                        actionHelp = this.ovaleSpellBook.GetSpellName(this.actionId) || "Unknown spell";
                    } else if (this.actionType == "value") {
                        actionHelp = (this.value < INFINITY) && tostring(this.value) || "infinity";
                    } else {
                        actionHelp = format("%s %s", this.actionType, tostring(this.actionId));
                    }
                }
                GameTooltip.AddLine(actionHelp, 0.5, 1, 0.75);
            }
            if (this.HasScriptControls()) {
                GameTooltip.AddLine(L["Cliquer pour afficher/cacher les options"], 1, 1, 1);
            }
            GameTooltip.Show();
        }
    }
    OvaleIcon_OnLeave() {
        if (this.help || this.HasScriptControls()) {
            GameTooltip.Hide();
        }
    }
    SetPoint(anchor: UIPosition, reference: UIFrame, refAnchor: UIPosition, x:number, y: number) {
        this.frame.SetPoint(anchor, reference, refAnchor, x, y);
    }

    Show(){
        this.frame.Show();
    }

    Hide() {
        this.frame.Hide();
    }

    SetScale(scale: number) {
        this.frame.SetScale(scale);
    }

    EnableMouse(enabled: boolean){
        this.frame.EnableMouse(enabled);
    }
}
